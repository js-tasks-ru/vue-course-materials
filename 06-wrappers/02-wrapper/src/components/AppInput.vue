<template>
  <!--  input-group_icon input-group_icon-left form-control_rounded form-control_sm -->
  <div
    class="input-group"
    :class="{
      'input-group_icon': hasIcon(),
      'input-group_icon-left': hasIcon(),
    }"
  >
    <slot name="left-icon" />
    <input
      class="form-control"
      :value="value"
      v-bind="$attrs"
      v-on="listeners"
    />
  </div>
</template>

<script>
export default {
  name: 'AppInput',

  inheritAttrs: false,

  model: {
    prop: 'value',
    event: 'input',
  },

  props: {
    value: String,
  },

  computed: {
    listeners() {
      return {
        ...this.$listeners,
        input: ($event) => {
          this.$emit('input', $event.target.value);
        },
      };
    },
  },

  methods: {
    hasIcon() {
      return !!this.$slots['left-icon'];
    },
  },
};
</script>

<style scoped>
.form-control {
  padding: 12px 16px;
  height: 52px;
  border-radius: 8px;
  border: 2px solid var(--blue-light);
  font-family: 'Nunito', sans-serif;
  font-weight: 600;
  font-size: 20px;
  line-height: 28px;
  color: var(--body-color);
  transition: 0.2s all;
  background-color: var(--white);
  outline: none;
  box-shadow: none;
}

.form-control::placeholder {
  font-weight: 400;
  color: var(--blue-2);
}

.form-control:focus {
  border-color: var(--blue);
}

textarea.form-control {
  width: 100%;
  min-height: 211px;
}

.form-control.form-control_rounded {
  border-radius: 26px;
}

.form-control.form-control_sm.form-control_rounded {
  border-radius: 22px;
}

.form-control.form-control_sm {
  padding: 8px 16px;
  height: 44px;
  border-radius: 4px;
}

.input-group {
  position: relative;
}

.input-group .form-control {
  width: 100%;
}

.input-group.input-group_icon .form-control {
  padding-left: 50px;
}

.input-group.input-group_icon .icon {
  position: absolute;
  top: 50%;
  transform: translate(0, -50%);
}

.input-group.input-group_icon.input-group_icon-left .icon {
  left: 16px;
}

.input-group.input-group_icon.input-group_icon-right .icon {
  right: 16px;
}
</style>
