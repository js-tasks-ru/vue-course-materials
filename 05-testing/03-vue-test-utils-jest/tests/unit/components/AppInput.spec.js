import { shallowMount } from '@vue/test-utils';
import AppInput from '@/components/AppInput';

// test('AppInput should be defined', () => {
//   expect(AppInput).toBeDefined();
// });

describe('AppInput', () => {
  it('should be defined', () => {
    expect(AppInput).toBeDefined();
  });

  it('should render input.input-group inside .form-control', async () => {
    const wrapper = shallowMount(AppInput);
    const inputGroup = wrapper.find('.input-group');
    const formControl = inputGroup.find('input.form-control');
    expect(inputGroup.exists()).toBe(true);
    expect(formControl.exists()).toBe(true);
  });

  it('should not have .form-control_rounded on input by default', async () => {
    const wrapper = shallowMount(AppInput);
    expect(wrapper.find('input').classes('form-control_rounded')).toBe(false);
  });

  it('should have .form-control_rounded on input with rounded', async () => {
    const wrapper = shallowMount(AppInput, { propsData: { rounded: true } });
    expect(wrapper.find('input').classes('form-control_rounded')).toBe(true);
  });

  it('should have toggle .form-control_rounded on input by rounded', async () => {
    const wrapper = shallowMount(AppInput);
    const input = wrapper.find('input');
    expect(input.classes('form-control_rounded')).toBe(false);

    await wrapper.setProps({ rounded: true });
    // await wrapper.vm.$nextTick();
    expect(input.classes('form-control_rounded')).toBe(true);
  });

  it('should have current value', async () => {
    const value = 'SameValue';
    const wrapper = shallowMount(AppInput, { propsData: { value } });
    expect(wrapper.find('input').element.value).toBe(value);
  });

  it.each(['password', 'num', 'range', 'date', 'time', 'email', 'tel'])(
    'should render input type=(%s)',
    (type) => {
      const wrapper = shallowMount(AppInput, { attrs: { type } });
      expect(wrapper.find('.form-control').attributes('type')).toBe(type);
    },
  );

  it('should emit input inputted value on input', async () => {
    const value = 'TEST_VALUE';
    const wrapper = shallowMount(AppInput);
    // wrapper.find('.form-control').element.value = value;
    // await wrapper.find('.form-control').trigger('input');
    await wrapper.find('.form-control').setValue(value);

    expect(wrapper.emitted().input).toBeTruthy();
    expect(wrapper.emitted().input).toHaveLength(1);
    expect(wrapper.emitted().input[0]).toEqual([value]);
  });

  it('should handle change event', async () => {
    const value = 'TEST_VALUE';
    const handler = jest.fn();
    const wrapper = shallowMount(AppInput, {
      listeners: {
        change: (event) => handler(event.target.value),
      },
    });

    await wrapper.find('.form-control').setValue(value);
    await wrapper.find('.form-control').trigger('change');

    expect(handler).toHaveBeenCalled();
    expect(handler).toHaveBeenCalledWith(value);
  });

  it('should match snapshot', async () => {
    const wrapper = shallowMount(AppInput);
    expect(wrapper).toMatchSnapshot();
  });
});
