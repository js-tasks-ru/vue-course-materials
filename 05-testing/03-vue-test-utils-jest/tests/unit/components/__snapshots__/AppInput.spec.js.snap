// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInput should match snapshot 1`] = `<div class="input-group"><input class="form-control"></div>`;
